### DISCUSSION TITLE
How to include optional dependencies that are instrument specific?

### DISCUSSION THREAD

**Discussion by emlynjdavies** [SINTEF/pyopia#14] [General]
i.e fftpack is needed for holography, but not UVP or SilCam

@nepstad I think you had some thoughts about optional dependencies in setuptools?

**Comment by emlynjdavies** [comment]
This is done in Xarray: https://docs.xarray.dev/en/stable/getting-started-guide/installing.html#id1

How could we implement something like this in PyOpia if needed? We would like to keep the complexity of dependencies to the minimum needed. i.e. if we wanted to build a light system to just process fast and give us size distributions (and no classification), we could skip installing tensorflow, for example.

Do you have any experience / ideas on this @dbuscombe-usgs ?

**Comment by ddbuscombe** [comment]
I have some thoughts on this, but will need a few days to compose them.

For clarification, you want different 'profiles' that install subset of dependencies depending on the user's needs? Also, is tensorflow-lite planned to be used for mobile/edge devices, or regular tensorflow?

**Reply by emlynjdavies** [reply]
great! Thanks!

Yes: "different 'profiles' that install subset of dependencies."

We are trying to make the pipeline modular, so that it should be possible to build a tensorflow-light version that could replace the example Classification class if people had a need for that. In this case the idea is that they would define a custom classification class that is called by Pipeline: https://pyopia.readthedocs.io/en/holo/pyopia.pipeline.html#pyopia.pipeline.Pipeline
and as long as it uses the same inputs and outputs as the standard Classification class, then it should work:
https://github.com/SINTEF/pyopia/blob/main/pyopia/classify.py
(note: the current classification class is just a copy-paste from an old PySilCam classifier, so likely needs some modernization at some point and this module also needs better documentation so people know what inputs and outputs they need to have to be compatible)

**Comment by emlynjdavies** [comment]
This can be done by added [options.extras_require] into setup.cfg if we merge #108 That would then eventually allow optional dependencies that would be installed with, for example:
```pip install pyopia[plotting]```

the the contents of setup.cfg would have something like:
```
[options.extras_require]
plotting = 
    matplotlib
    cmocean
    plotly
```

**Comment by emlynjdavies** [comment]
See #140
